File: Sources/HedosKernel/Artifacts/Artifact.swift
import Foundation

public struct Artifact: Codable, Hashable, Sendable, Identifiable {
    public let id: String
    public var path: String
    public var contentHash: String
    public var previewPath: String?
    public var model: String
    public var modelID: String
    public var runtime: String
    public var capability: Capability
    public var params: JSONValue
    public var createdAt: Date
    public var durationMs: Int
    public var jobID: String
    public var sessionID: String?

    public init(
        id: String,
        path: String,
        contentHash: String,
        previewPath: String? = nil,
        model: String,
        modelID: String,
        runtime: String,
        capability: Capability,
        params: JSONValue,
        createdAt: Date,
        durationMs: Int,
        jobID: String,
        sessionID: String? = nil
    ) {
        self.id = id
        self.path = path
        self.contentHash = contentHash
        self.previewPath = previewPath
        self.model = model
        self.modelID = modelID
        self.runtime = runtime
        self.capability = capability
        self.params = params
        self.createdAt = createdAt
        self.durationMs = durationMs
        self.jobID = jobID
        self.sessionID = sessionID
    }

    enum CodingKeys: String, CodingKey {
        case id, contentHash, model, modelID, runtime, capability, params
        case createdAt, durationMs, jobID, sessionID
        case path = "artifact"
        case previewPath = "preview"
    }
}

public struct ArtifactDraft: Sendable {
    public var data: Data
    public var fileExtension: String
    public var preview: Data?
    public var model: String
    public var modelID: String
    public var runtime: String
    public var capability: Capability
    public var params: JSONValue
    public var jobID: String
    public var durationMs: Int
    public var sessionID: String?

    public init(
        data: Data,
        fileExtension: String,
        preview: Data? = nil,
        model: String,
        modelID: String,
        runtime: String,
        capability: Capability,
        params: JSONValue,
        jobID: String,
        durationMs: Int,
        sessionID: String? = nil
    ) {
        self.data = data
        self.fileExtension = fileExtension
        self.preview = preview
        self.model = model
        self.modelID = modelID
        self.runtime = runtime
        self.capability = capability
        self.params = params
        self.jobID = jobID
        self.durationMs = durationMs
        self.sessionID = sessionID
    }
}

File: kernel/src/artifacts/artifact.rs
//! The stored `Artifact` record and the `ArtifactDraft` a writer submits.

use serde::{Deserialize, Serialize};

use crate::records::{Capability, JsonValue};

/// A stored generated output: where its bytes live, a content hash, provenance
/// (model/runtime/capability/params), timing, and ownership (job, session).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Artifact {
    pub id: String,
    /// The output file path, relative to the store root.
    pub path: String,
    /// SHA-256 (hex) of the output bytes.
    pub content_hash: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub preview_path: Option<String>,
    /// The model's display name.
    pub model: String,
    /// The model's registry id.
    pub model_id: String,
    pub runtime: String,
    pub capability: Capability,
    pub params: JsonValue,
    /// Creation time, epoch milliseconds.
    pub created_at: i64,
    pub duration_ms: i64,
    pub job_id: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub session_id: Option<String>,
}

/// Everything needed to store one output. The store fills in the id, path,
/// content hash, preview path, and creation time.
#[derive(Debug, Clone)]
pub struct ArtifactDraft {
    pub data: Vec<u8>,
    pub file_extension: String,
    pub preview: Option<Vec<u8>>,
    pub model: String,
    pub model_id: String,
    pub runtime: String,
    pub capability: Capability,
    pub params: JsonValue,
    pub job_id: String,
    pub duration_ms: i64,
    pub session_id: Option<String>,
}
